:sparkles: Add namespace scoped mode

Sunnatillo opened the pull request:
<!-- please add a icon to the title of this PR and delete this line and similar ones -->
<!-- the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other) -->

This PR introduces namespace scoped mode for BMO. We can specify namespaces to watch for BMO by "WATCH_NAMESPACE" variable. 

- Makes main branch  e2e tests to run in namespaced scope.

Please read "namespace-scoped-mode.md" document before reviewing. 


**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

tuminoid commented:
/cc @tuminoid 
I'll be taking a look shortly.

lentzi90 commented on docs/namespace-scoped-setup.md:
```diff
@@ -0,0 +1,325 @@
+  kind: RoleBinding
+  metadata:
+    name: baremetal-operator-manager-rolebinding
+    namespace: bar
```
Is it possible to skip the namespace for the Role and RoleBinding here and just add it through kustomize/kubectl? That was what I had in mind, i.e. we could provide these manifests and they can then easily be applied in any namespace that should be watched, like this:
```bash
kubectl -n foo -f bmo-namespaced-rbac.yaml
kubectl -n bar -f bmo-namespaced-rbac.yaml
kubectl -n baz -f bmo-namespaced-rbac.yaml
```

Sunnatillo replied:
Yes, it is possible. This would be a manual work. 
By doing this we will not use kustomize plugin.

lentzi90 replied:
Good! The plugin is really useful I think, but it is nice to have an easy, manual way of doing it also. Some will prefer it that way for one reason or another :slightly_smiling_face:

lentzi90 reviewed (commented):
Good progress @Sunnatillo !
I would honestly prefer to do it without the plugin first, just to get a feeling for how the manual way works. The plugin also has the limitation that it is mainly useful during initial setup. I would like to see how to work with the manifests in a dynamic environment.

For this reason I would also like to avoid making the e2e namespaces static. Perhaps it would turn into a mess, but this is the first thing I would like to try. I.e. apply the extra RBAC in the namespace as part of the test and modify BMO to include it. Modifying BMO will probably be the problematic part as that will disrupt the controller once for every test... I have probably asked this before, but is there a way to make the controller select namespaces based on labels instead? That would solve this part neatly!

tuminoid commented on config/overlays/e2e/namespace-scoped-resources.yaml:
we want only one role and one rolebinding here, without namespace, so they can be applied in a loop to each namespace.

tuminoid commented on config/overlays/e2e/roles-ns-annotator.yaml:
we don't need this

tuminoid commented on docs/namespace-scoped-setup.md:
update the documentation to the new simpler way of applying this, and also document the limitations and recovery if you screw up with the namespace.

tuminoid commented on hack/tools/kustomize-namespace-annotator/README.md:
none of this is needed as discussed

tuminoid commented on tools/deploy.sh:
```diff
@@ -223,8 +223,9 @@ if [[ "${DEPLOY_BMO}" == "true" ]]; then
+
     # shellcheck disable=SC2086
-    ${KUSTOMIZE} build "${TEMP_BMO_OVERLAY}" | kubectl apply ${KUBECTL_ARGS} -f -
+    ${KUSTOMIZE} build --enable-alpha-plugins --enable-exec "${TEMP_BMO_OVERLAY}" | kubectl apply ${KUBECTL_ARGS} -f -
```
plugin not needed

tuminoid reviewed (commented):
Summarizing as rough review comments what we discussed as a group in the meeting how this could be simplified.

tuminoid commented on config/overlays/namespaced/namespaced-manager-patch.yaml:
```diff
@@ -0,0 +1,13 @@
+      containers:
+      - env:
+        - name: WATCH_NAMESPACE
+          value: basic-ops,external-inspection,externally-provisioned,inspection,live-iso-ops,provisioning-ops,re-inspection,automated-cleaning
```
This should probably be foo,bar,quo

tuminoid commented on config/overlays/namespaced/kustomization.yaml:
```diff
@@ -0,0 +1,24 @@
+kind: Kustomization
+resources:
+- ../../base
+- ns-foo.yaml
```
We're still not getting rid of these copies?

tuminoid commented on config/overlays/e2e/kustomization.yaml:
```diff
@@ -1,11 +1,19 @@
-namespace: baremetal-operator-system
 resources:
 - ../basic-auth_tls
+- ./namespaced/automated-cleaning-resources.yaml
```
What is wrong with a loop in e2e script that applies role/rolebinding to all relevant namespaces? We shouldn't force that in kustomize IMO?

@lentzi90 PTAL

lentzi90 replied:
I think we can do it two ways:
1. Loop in bash or go where you do something like
for ns in test_namespaces; kubectl create ns $ns && kubectl -n $ns apply -k namespaced; done
2. Declaratively define overlays for each and then pull them all in one for applying easily.

I think 2 is what you were sort of going for here, but with only one kustomization. We can reduce the amount of yaml a lot if you do separate overlays for each namespace. Like this:
```
├── basic-ops
│   ├── kustomization.yaml
│   └── namespace.yaml
├── live-iso
│   ├── kustomization.yaml
│   └── namespace.yaml
└── re-inspection
    ├── kustomization.yaml
    └── namespace.yaml
```

In each kustomization.yaml you then put something like this:

```yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
resources:
- ../../namespaced
- namespace.yaml

namespace: basic-ops
```

Now you don't have the full manifest but instead a kustomization that generates it when you need it. You can then put them all together in one kustomization:

```yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization
resources:
- basic-ops
- live-iso
- re-inspection
- ...
```

Sunnatillo replied:
I have done the changes you requested. But I have placed them inside component so it can be called for other tests when we include them.

Sunnatillo commented:
/retest

Sunnatillo commented on test/e2e/common.go:
```diff
@@ -218,6 +219,24 @@ func cleanup(ctx context.Context, clusterProxy framework.ClusterProxy, namespace
+	// This way there should be no risk of BMO getting stuck trying to progress
+	// and create HardwareDetails or similar, while the namespace is terminating.
+	DeleteBmhsInNamespace(ctx, clusterProxy.GetClient(), namespace.Name)
+	if !isNamespaced {
+		framework.DeleteNamespace(ctx, framework.DeleteNamespaceInput{
+			Deleter: clusterProxy.GetClient(),
+			Name:    namespace.Name,
+		})
+		WaitForNamespaceDeleted(ctx, WaitForNamespaceDeletedInput{
+			Getter:    clusterProxy.GetClient(),
+			Namespace: *namespace,
+		}, intervals...)
+	}
```
yes, we cannot delete namespace because of limitation in controller runtime. 
I will add explanation here.

Sunnatillo commented on test/e2e/common.go:
```diff
@@ -234,7 +253,12 @@ func WaitForBmhInPowerState(ctx context.Context, input WaitForBmhInPowerStateInp
 
 func BuildKustomizeManifest(source string) ([]byte, error) {
-	kustomizer := krusty.MakeKustomizer(krusty.MakeDefaultOptions())
+	kustomizeOptions := krusty.MakeDefaultOptions()
+	// Enable plugins
+	kustomizeOptions.PluginConfig = kustomizetypes.MakePluginConfig(
+		kustomizetypes.PluginRestrictionsNone,
+		kustomizetypes.BploUseStaticallyLinked)
+	kustomizer := krusty.MakeKustomizer(kustomizeOptions)
```
ohh, yes, this is leftover from kustomize plugin implementation.

Sunnatillo commented on test/e2e/inspection_test.go:
```diff
@@ -51,6 +68,23 @@ var _ = Describe("Inspection", Label("required", "inspection"), func() {
+		err = clusterProxy.GetClient().Delete(ctx, &bmh)
+		Expect(err).NotTo(HaveOccurred())
+
+		By("Waiting for the BMH to be deleted")
```
We have discussed this. We run parallel tests. Those tests when reach AfterEach another test will already start, there is a time bmh is not deleted in first test and second test starts and bmh fails to register. 
Thus we decided to keep Delete BMH step in tests.

Sunnatillo commented on config/overlays/e2e/kustomization.yaml:
```diff
@@ -1,11 +1,14 @@
 - ../basic-auth_tls
 
+components:
+  - ../../components/namespaced
```
I have implemented removal of ClusterRole and CLusterRolebinging as a component. Creation of Roles and RoleBindings are put in e2e overlay.

lentzi90 commented on config/overlays/namespaced/kustomization.yaml:
```diff
@@ -0,0 +1,24 @@
+resources:
+- ../../base
+- ns-foo.yaml
+- ns-bar.yaml
+- ns-quo.yaml
+
+patches:
+# Remove ClusterRoleBinding (namaspace scoped mode)
+- patch: |
+    apiVersion: rbac.authorization.k8s.io/v1
+    kind: ClusterRoleBinding
```
As an overlay, I think it is better to show how to use the component, but we already have the e2e overlay that can work as example to show that. I am fine with keeping a plain manifest example also, but then there isn't really much point in having the kustomization here.
I would suggest just removing this. We can document the whole process in the metal3-docs instead.

lentzi90 commented on test/e2e/automated_cleaning_test.go:
```diff
@@ -198,11 +198,28 @@ var _ = Describe("Automated cleaning", Label("required", "automated-cleaning"),
 		DumpResources(ctx, e2eConfig, clusterProxy, path.Join(artifactFolder, specName))
 		if !skipCleanup {
-			cleanup(ctx, clusterProxy, namespace, cancelWatches, e2eConfig.GetIntervals("default", "wait-namespace-deleted")...)
+			namespaced := e2eConfig.GetVariable("NAMESPACE_SCOPED")
+			isNamespaced, err := strconv.ParseBool(namespaced)
+			if err != nil {
+				// if error it sets variable to false
+				isNamespaced = false
+			}
```
```suggestion
			isNamespaced := e2eConfig.GetBoolVariable("NAMESPACE_SCOPED")
```

lentzi90 commented on test/e2e/inspection_test.go:
```diff
@@ -26,12 +27,28 @@ var _ = Describe("Inspection", Label("required", "inspection"), func() {
 	BeforeEach(func() {
-		namespace, cancelWatches = framework.CreateNamespaceAndWatchEvents(ctx, framework.CreateNamespaceAndWatchEventsInput{
+
+		namespaced := e2eConfig.GetVariable("NAMESPACE_SCOPED")
+		isNamespaced, err := strconv.ParseBool(namespaced)
+		if err != nil {
+			// if error it sets variable to false
+			isNamespaced = false
+		}
```
```suggestion
		isNamespaced := e2eConfig.GetBoolVariable("NAMESPACE_SCOPED")
```

lentzi90 commented on test/e2e/inspection_test.go:
```diff
@@ -125,12 +164,29 @@ var _ = Describe("Inspection", Label("required", "inspection"), func() {
 		DumpResources(ctx, e2eConfig, clusterProxy, path.Join(artifactFolder, specName))
 		if !skipCleanup {
-			cleanup(ctx, clusterProxy, namespace, cancelWatches, e2eConfig.GetIntervals("default", "wait-namespace-deleted")...)
+			namespaced := e2eConfig.GetVariable("NAMESPACE_SCOPED")
+			isNamespaced, err := strconv.ParseBool(namespaced)
+			if err != nil {
+				// if error it sets variable to false
+				isNamespaced = false
+			}
```
```suggestion
			isNamespaced := e2eConfig.GetBoolVariable("NAMESPACE_SCOPED")
```

lentzi90 commented on test/e2e/live_iso_test.go:
```diff
@@ -136,12 +136,29 @@ var _ = Describe("Live-ISO", Label("required", "live-iso"), func() {
 		DumpResources(ctx, e2eConfig, clusterProxy, path.Join(artifactFolder, specName))
 		if !skipCleanup {
-			cleanup(ctx, clusterProxy, namespace, cancelWatches, e2eConfig.GetIntervals("default", "wait-namespace-deleted")...)
+			namespaced := e2eConfig.GetVariable("NAMESPACE_SCOPED")
+			isNamespaced, err := strconv.ParseBool(namespaced)
+			if err != nil {
+				// if error it sets variable to false
+				isNamespaced = false
+			}
```
```suggestion
			isNamespaced := e2eConfig.GetBoolVariable("NAMESPACE_SCOPED")
```

lentzi90 commented on test/e2e/provisioning_and_annotation_test.go:
```diff
@@ -253,12 +253,29 @@ var _ = Describe("Provision, detach, recreate from status and deprovision", Labe
 			DumpResources(ctx, e2eConfig, clusterProxy, path.Join(artifactFolder, specName))
 			if !skipCleanup {
-				cleanup(ctx, clusterProxy, namespace, cancelWatches, e2eConfig.GetIntervals("default", "wait-namespace-deleted")...)
+				namespaced := e2eConfig.GetVariable("NAMESPACE_SCOPED")
+				isNamespaced, err := strconv.ParseBool(namespaced)
+				if err != nil {
+					// if error it sets variable to false
+					isNamespaced = false
+				}
```
```suggestion
				isNamespaced := e2eConfig.GetBoolVariable("NAMESPACE_SCOPED")
```

lentzi90 commented on test/e2e/re_inspection_test.go:
```diff
@@ -107,12 +107,29 @@ var _ = Describe("Re-Inspection", Label("required", "re-inspection"), func() {
 		DumpResources(ctx, e2eConfig, clusterProxy, path.Join(artifactFolder, specName))
 		if !skipCleanup {
-			cleanup(ctx, clusterProxy, namespace, cancelWatches, e2eConfig.GetIntervals("default", "wait-namespace-deleted")...)
+			namespaced := e2eConfig.GetVariable("NAMESPACE_SCOPED")
+			isNamespaced, err := strconv.ParseBool(namespaced)
+			if err != nil {
+				// if error it sets variable to false
+				isNamespaced = false
+			}
```
```suggestion
			isNamespaced := e2eConfig.GetBoolVariable("NAMESPACE_SCOPED")
```

lentzi90 commented on docs/namespace-scoped-setup.md:
```diff
@@ -0,0 +1,206 @@
+
+**Example: Configuring BMO to watch its own deployed namespace:**
+
+To configure the operator to automatically watch the same namespace it's
+deployed in, modify the `config/base/manager.yaml` file. Add the following
```
The example below is watching foo,bar and quo.
```suggestion
To configure the operator to automatically watch the same namespaces foo,
bar and quo, modify the `config/base/manager.yaml` file. Add the following
```

Sunnatillo replied:
Done as well

Sunnatillo commented:
/retest

Sunnatillo commented:
/retest

fmuyassarov reviewed (approved):
I didn't go through the e2e part and only kustomize related changes and it looks good to me. Good work @Sunnatillo .

Since we are here, I think as a follow up it would be good to revisit the whole structure of the kustomize directory. Because, it seems to me that we are not following the actual structure of the kustomize and there are some bits that can be reused and reduce the duplication and complexity. But of course, it has nothing to do with this PR.

tuminoid commented:
I'll find time to review this next week, holding meanwhile.
/hold

Sunnatillo commented on docs/namespace-scoped-setup.md:
```diff
@@ -0,0 +1,206 @@
+## 2\. Create Roles and RoleBindings for each namespaces
+
+When BMO operates in a namespace-scoped mode, you need to create `Roles` and
+`Rolebindings` for each namespaces that you added in WATCH_NAMESPACE in env
```
fixed

Sunnatillo commented on config/components/namespace-scoped/kustomization.yaml:
```diff
@@ -0,0 +1,18 @@
+    metadata:
+      name: baremetal-operator-manager-rolebinding
+    $patch: delete
+# Remove ClusterRole (namaspace scoped mode)
```
fixed

Sunnatillo commented on config/components/namespace-scoped/kustomization.yaml:
```diff
@@ -0,0 +1,18 @@
+    metadata:
+      name: baremetal-operator-manager-rolebinding
+    $patch: delete
+# Remove ClusterRole (namaspace scoped mode)
```
fixed

Sunnatillo commented:
> Fix the typos in readme, then let's unhold. There are basically only e2e fixes, and the actual change in the controller is near zero.

Addressed the comments. if LGTY we can merge this

Sunnatillo commented:
/retest 
Live iso is broken in CI. Not related to the changes

Sunnatillo commented:
/retest

lentzi90 commented:
I think the test is failing because an inspection test was added in https://github.com/metal3-io/baremetal-operator/pull/2670. You will need to adapt

Rozzii reviewed (commented):
/lgtm

tuminoid reviewed (commented):
/lgtm
/retitle :sparkles: Add namespace scoped mode

Let's merge this, but prepare to revert it if something unexpectedly breaks.
